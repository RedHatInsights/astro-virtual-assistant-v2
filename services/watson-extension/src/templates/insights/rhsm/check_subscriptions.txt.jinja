{%- if subs_info -%}
{%- if subs_info|length == 1 -%}
You have {{subs_info[0].number}} {{subs_info[0].category}} subscriptions
{%- else -%}
You have:

{% for sub_info in subs_info -%}
{% if sub_info.category == "active" -%}
* {{ sub_info.number }} active subscriptions
{% elif sub_info.category == "expiring" -%}
* {{ sub_info.number }} subscriptions that are expiring soon
{% elif sub_info.category == "expired" -%}
* {{ sub_info.number }} that have already expired
{%- endif -%}
{%- endfor -%}
{%- endif -%}
{%- else -%}
You don't have any subscriptions
{%- endif -%}
