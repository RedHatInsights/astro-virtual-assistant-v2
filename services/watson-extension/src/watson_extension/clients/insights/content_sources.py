import abc
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import ContentSourcesURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class GetPopularRepositoriesResponse:
    data: Optional[List[dict]]


@dataclass
class BulkCreateRepositoriesInfo:
    suggested_name: str
    distribution_arch: str
    distribution_versions: List[str]
    gpg_key: str
    metadata_verification: bool
    snapshot: bool
    url: str


@dataclass
class RepositoriesBulkCreateResponse:
    response: Optional[str] = None


class ContentSourcesClient(abc.ABC):
    @abc.abstractmethod
    async def get_popular_repositories(self) -> GetPopularRepositoriesResponse: ...

    @abc.abstractmethod
    async def repositories_bulk_create(
        self,
        repository_info: List[BulkCreateRepositoriesInfo],
    ) -> RepositoriesBulkCreateResponse: ...


class ContentSourcesClientHttp(ContentSourcesClient):
    def __init__(
        self,
        content_sources_url: injector.Inject[ContentSourcesURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.content_sources_url = content_sources_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_popular_repositories(self) -> GetPopularRepositoriesResponse:
        request = "/api/content-sources/v1/popular_repositories/?offset=0&limit=20"
        response = await self.platform_request.get(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        return GetPopularRepositoriesResponse(data=content["data"])

    async def repositories_bulk_create(
        self,
        repository_info: List[BulkCreateRepositoriesInfo],
    ) -> RepositoriesBulkCreateResponse:
        headers = {"Content-Type": "application/json"}

        request = "/api/content-sources/v1.0/repositories/bulk_create/"
        response = await self.platform_request.post(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=repository_info,
            headers=headers,
        )
        status = response.status

        content = await response.json()

        errors = None
        if "errors" in content:
            errors = content["errors"]

        repositories_response = None
        if status >= 400 and any(
            "already belongs" in error["detail"] for error in errors
        ):
            repositories_response = "already_enabled"
        elif status == 201:
            repositories_response = "enabled"
        else:
            repositories_response = "errors"

        return RepositoriesBulkCreateResponse(response=repositories_response)
