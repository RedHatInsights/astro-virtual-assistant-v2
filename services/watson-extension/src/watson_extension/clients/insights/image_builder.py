import abc
from dataclasses import dataclass
from typing import Optional

import injector

from watson_extension.clients import ContentSourcesURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class EnableCustomRepositoriesResponse:
    response: Optional[str] = None


class ImageBuilderClient(abc.ABC):
    @abc.abstractmethod
    async def enable_custom_repositories(
        self, version: str
    ) -> EnableCustomRepositoriesResponse: ...


class ImageBuilderClientHttp(ImageBuilderClient):
    def __init__(
        self,
        content_sources_url: injector.Inject[ContentSourcesURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.content_sources_url = content_sources_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def enable_custom_repositories(
        self,
        version: str,
    ) -> EnableCustomRepositoriesResponse:
        request = "/api/content-sources/v1/popular_repositories/?offset=0&limit=20"
        response = await self.platform_request.get(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        repository = None
        for repo in content["data"]:
            if repo["suggested_name"] and repo["suggested_name"].startswith(version):
                repository = repo
                break

        headers = {"Content-Type": "application/json"}
        formatted = [
            {
                "name": repository["suggested_name"],
                "distribution_arch": repository["distribution_arch"],
                "distribution_versions": repository["distribution_versions"],
                "gpg_key": repository["gpg_key"],
                "metadata_verification": repository["metadata_verification"],
                "snapshot": False,
                "url": repository["url"],
            }
        ]

        request = "/api/content-sources/v1.0/repositories/bulk_create/"
        response = await self.platform_request.post(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=formatted,
            headers=headers,
        )
        status = response.status

        content = await response.json()

        errors = None
        if "errors" in content:
            errors = content["errors"]

        repositories_response = None
        if status >= 400 and any(
            "already belongs" in error["detail"] for error in errors
        ):
            repositories_response = "already_enabled"
        elif status == 201:
            repositories_response = "enabled"
        else:
            repositories_response = "errors"

        return EnableCustomRepositoriesResponse(response=repositories_response)
