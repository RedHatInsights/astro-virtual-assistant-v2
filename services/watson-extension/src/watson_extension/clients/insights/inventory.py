import abc
import enum
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import RHSMURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class ActivationKeyResponse:
    ok: bool
    message: Optional[str]


class InventoryClient(abc.ABC):
    @abc.abstractmethod
    async def create_activation_key(self, name: str): ...


class InventoryClientHttp(InventoryClient):
    def __init__(
        self,
        rhsm_url: injector.Inject[RHSMURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.rhsm_url = rhsm_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def create_activation_key(self, name: str):
        body = {"name": name, "role": "", "serviceLevel": "", "usage": ""}
        response = await self.platform_request.post(
            self.rhsm_url,
            "/api/rhsm/v2/activation_keys",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=body,
        )
        response.raise_for_status()

        response_json = await response.json()

        if not response.ok:
            return ActivationKeyResponse(
                ok=response.ok,
                message=response_json.get("error", {}).get(
                    "message", "Unknown error"
                )
            )
        return ActivationKeyResponse(ok=response.ok, message=None)
