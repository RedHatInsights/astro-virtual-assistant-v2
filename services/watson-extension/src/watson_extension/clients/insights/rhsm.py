import abc
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import RhsmURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class SubscriptionInfo:
    number: int
    category: str


class RhsmClient(abc.ABC):
    @abc.abstractmethod
    async def check_subscriptions(
        self, category: Optional[str]
    ) -> List[SubscriptionInfo]: ...


class RhsmClientHttp(RhsmClient):
    def __init__(
        self,
        rhsm_url: injector.Inject[RhsmURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.rhsm_url = rhsm_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def check_subscriptions(
        self,
        category: Optional[str],
    ) -> List[SubscriptionInfo]:
        request = "/api/rhsm/v2/products/status"
        response = await self.platform_request.get(
            self.rhsm_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()
        content_body = content.get("body")

        subscriptions_info = []
        if content_body:
            if category:
                category_text = category
                if category == "expiringSoon":
                    category_text = "expiring"

                subs_category_count = content_body.get(category)
                subscriptions_info.append(
                    SubscriptionInfo(number=subs_category_count, category=category_text)
                )
            else:
                subscriptions_info = []
                if content_body.get("active"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["active"], category="active"
                        )
                    )
                if content_body.get("expiringSoon"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["expiringSoon"], category="expiring"
                        )
                    )
                if content_body.get("expired"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["expired"], category="expired"
                        )
                    )

        return subscriptions_info
