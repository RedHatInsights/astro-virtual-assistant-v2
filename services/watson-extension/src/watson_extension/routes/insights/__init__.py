from quart import Blueprint
from . import advisor, inventory, vulnerability, content_sources, rhsm

blueprint = Blueprint("insights", __name__, url_prefix="/insights")

blueprint.register_blueprint(advisor.blueprint)
blueprint.register_blueprint(inventory.blueprint)
blueprint.register_blueprint(vulnerability.blueprint)
blueprint.register_blueprint(content_sources.blueprint)
blueprint.register_blueprint(rhsm.blueprint)
