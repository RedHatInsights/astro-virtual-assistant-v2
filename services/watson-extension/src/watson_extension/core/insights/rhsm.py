import injector
import enum
from typing import Optional

from watson_extension.clients.insights.rhsm import RhsmClient


class SubscriptionsCategory(enum.Enum):
    ACTIVE = "active"
    EXPIRED = "expired"
    EXPIRINGSOON = "expiringSoon"


class RhsmCore:
    def __init__(self, rhsm_client: injector.Inject[RhsmClient]):
        self.rhsm_client = rhsm_client

    async def check_subscriptions(
        self, category: Optional[SubscriptionsCategory] = None
    ):
        if category:
            return await self.rhsm_client.check_subscriptions(category.value)
        return await self.rhsm_client.check_subscriptions(None)
