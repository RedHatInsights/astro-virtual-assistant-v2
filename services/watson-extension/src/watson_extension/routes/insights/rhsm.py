import injector
from typing import Optional
from pydantic import BaseModel

from quart import Blueprint, render_template
from quart_schema import validate_response, validate_querystring, document_headers

from watson_extension.core.insights.rhsm import (
    SubscriptionsCategory,
    RhsmCore,
)
from watson_extension.routes import RHSessionIdHeader

blueprint = Blueprint("rhsm", __name__, url_prefix="/rhsm")


class SubscriptionsCategoryQuery(BaseModel):
    category: Optional[SubscriptionsCategory] = None


class CheckSubscriptionsResponse(BaseModel):
    response: str


@blueprint.get("/check_subscriptions")
@validate_querystring(SubscriptionsCategoryQuery)
@validate_response(CheckSubscriptionsResponse)
@document_headers(RHSessionIdHeader)
async def check_subscriptions(
    query_args: SubscriptionsCategoryQuery,
    rhsm_service: injector.Inject[RhsmCore],
) -> CheckSubscriptionsResponse:
    subscriptions_info = await rhsm_service.check_subscriptions(query_args.category)
    return CheckSubscriptionsResponse(
        response=await render_template(
            "insights/rhsm/check_subscriptions.txt.jinja",
            subs_info=subscriptions_info,
        )
    )
