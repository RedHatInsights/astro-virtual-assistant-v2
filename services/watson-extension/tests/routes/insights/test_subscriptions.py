from unittest.mock import MagicMock

import injector
import pytest
from quart.typing import TestClientProtocol

from watson_extension.clients.insights.rhsm import SubscriptionInfo, RhsmClient
from ..common import app_with_blueprint

from watson_extension.routes.insights.rhsm import blueprint
from ... import async_value


@pytest.fixture
async def rhsm_client() -> MagicMock:
    return MagicMock(RhsmClient)


@pytest.fixture
async def test_client(rhsm_client) -> TestClientProtocol:
    def injector_binder(binder: injector.Binder):
        binder.bind(RhsmClient, rhsm_client)

    return app_with_blueprint(blueprint, injector_binder).test_client()


async def test_check_subscriptions_all(test_client, rhsm_client, snapshot) -> None:
    rhsm_client.check_subscriptions = MagicMock(
        return_value=async_value(
            [
                SubscriptionInfo(
                    number="200",
                    category="active",
                ),
                SubscriptionInfo(
                    number="25",
                    category="expiring",
                ),
                SubscriptionInfo(
                    number="30",
                    category="expired",
                ),
            ]
        )
    )

    response = await test_client.get(
        "/rhsm/check_subscriptions",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot


async def test_check_subscriptions_one(test_client, rhsm_client, snapshot) -> None:
    rhsm_client.check_subscriptions = MagicMock(
        return_value=async_value(
            [
                SubscriptionInfo(
                    number="10",
                    category="expiring",
                ),
            ]
        )
    )

    response = await test_client.get(
        "/rhsm/check_subscriptions?category=expiringSoon",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot


async def test_check_subscriptions_none(test_client, rhsm_client, snapshot) -> None:
    rhsm_client.check_subscriptions = MagicMock(return_value=async_value([]))

    response = await test_client.get(
        "/rhsm/check_subscriptions",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot
