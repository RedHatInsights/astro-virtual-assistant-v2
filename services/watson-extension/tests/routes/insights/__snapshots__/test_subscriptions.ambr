# serializer version: 1
# name: test_check_subscriptions_all
  '''
  You have:
  
  * 200 active subscriptions
  * 25 subscriptions that are expiring soon
  * 30 that have already expired
  '''
# ---
# name: test_check_subscriptions_none
  "You don't have any subscriptions"
# ---
# name: test_check_subscriptions_one
  'You have 10 expiring subscriptions'
# ---
